nav2_map_server/include/nav2_map_server/occ_grid_loader.hpp

```diff
@@ -18,9 +18,9 @@
 #include <memory>
```

Comment:
What do you think about rolling the occ grid loader stuff into the MapIO class? 

I'm not sure whether its a good or bad idea. I'm thinking that MapIO should be dealing with interacting with the yaml / pgm / jpeg / png / ... files and loading them into memory. But maybe then the occ grid loader's role is to convert that data into the specific occupancy grid standard? But it looks like the MapIO function signatures already return a `nav_msgs::msg::OccupancyGrid` so it has to know about ROS / occupancy grids. 

Its unclear to me the separation of their roles. If we have to push the logic for occupancy grids into MapIO, then what specific role does the occ grid loader class server? I think the 2 options would be to either remove the occupancy grid logic from MapIO to make it more "pure" or put the occupancy grid loader code into MapIO. 

What do you think? Lets discuss before you do anything.

Reply:
That is the main question I asked myself after I moved all save/load logic into MapIO. 
The initial idea contributed by Intel - is to have `MapServer` abstract class and `OccGridLoader` derived interface with all OccupancyGrid-related logic. If I want to add new type of maps (e.g. OccupacyCircle type), it easily could be added just by making new derived `OccCircleLoader` class.

Let's return back to current change. From one hand, MapIO contains all map input-output logic related to OccupancyGrid type. Since all functions API from MapIO do contain OccupancyGrid parameters types, they can not be called directly from abstract savers/loaders without a wrapper (OccGridLoader - is one for loading). In this case the good choice is to make `MapSaver` to be an abstract class like MapServer and new `OccGridSaver` wrapper to be related to OccupancyGrid type saving methods from MapIO. Thus per each new type of map we need to add two wrapper classes (loader and saver) and one dynamic library. Seems to be complicated... And does dynamic library really needed in this case.

From other hand, we can not completely refuse from MapIO, since the first requirement of #1010 (create lib for file IO that can be used as a basis for executable) won't be fulfilled.

Thus, currently I've decided to remain `OccGridLoader` class as an OccupancyGrid wrapper (as it initially intended) and MapIO library for OccGrid inpiut-output. Possibly in the future `OccGridSaver` wrapper will be added to remain `MapSaver` to be an Occupancy-grid-independent abstract class. What do you think?

Reply:
Don't consider the non-occupancy grid types. We're not discussing adding that in the immediate or medium term future. When we would, it would make no sense to have a single package to deal with both occupancy grids and XYZ format (like a gradient map). The format of data and the mapping of information is specialized and we don't need to go crazy on trying to support everything in 1 package. Consider only the case of occupancy grids in some image-like format. That design was severely over-optimized and never had a clear application.

MapIO is for the basic IO for loading occupancy grid maps from file. Like you mention, it already makes use of the parameter types for it. Everything in this package is for occupancy grids and similar structured data (like the keep out maps and lanes, like this work is meant to be a predecessor for). Lets not get carried away.

Reply:
So, in this case, I think we can refuse the OccGridLoader and move its functionality inside MapServer class. This will greatly simplify the whole picture.

Reply:
OK, after looking at it, I agree. I thought there was more overlap with MapIO but I think you're right its mostly functional error checking / service code.

Reply:
Moved OccGridLoader functionality into MapServer class in https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7